src/Datadog.Trace.ClrProfiler.Managed/Emit/DynamicMethodBuilder.cs

```diff
@@ -42,6 +42,152 @@ internal static class DynamicMethodBuilder<TDelegate>
                     key.MethodGenericArguments));
         }
 
+        /// <summary>
+        /// Creates a simple <see cref="DynamicMethod"/> using <see cref="System.Reflection.Emit"/> that
+        /// calls a method with the specified name and parameter types.
+        /// </summary>
+        /// <param name="owningType">The <see cref="Type"/> that contains the method to call when the returned delegate is executed..</param>
+        /// <param name="methodName">The name of the method to call when the returned delegate is executed.</param>
+        /// <param name="returnType">Use method overload that matches the specified return owningType.</param>
+        /// <param name="parameterTypes">If not null, use method overload that matches the specified parameters.</param>
+        /// <param name="genericTypes">If not null, use method overload that has the same number of generic arguments.</param>
+        /// <returns>A <see cref="Delegate"/> that can be used to execute the dynamic method.</returns>
+        public static TDelegate CreateInstrumentedMethodDelegate(
+            Type owningType,
+            string methodName,
+            Type returnType,
+            Type[] parameterTypes,
+            Type[] genericTypes)
+        {
+            MethodInfo[] methods =
+                owningType.GetMethods(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Static);
+
+            MethodInfo candidate = null;
+            Type[] candidateParameterTypes = null;
+
+            for (ushort i = 0; i < methods.Length; i++)
+            {
+                if (methods[i].Name != methodName)
+                {
+                    continue;
+                }
+
+                // We don't know the return type in many cases
+                if (returnType != null)
+                {
+                    if (methods[i].ReturnType.AssemblyQualifiedName != returnType.AssemblyQualifiedName)
```

Comment:
We might be able to just use the Equals instance method to compare `methods[i].ReturnType` and `returnType`, unless there's a more compelling reason to do a String comparison.

Reply:
I think what we actually have to do here is use `IsAssignableTo`.